StackOverFlowError during incremental Java compilation with annotation processors

The following contains two issues, but I report them in a single issue due to being related to the same codebase. Feel free to create a separate issue for the compilation error part.

### Expected Behavior
1. When recompiling the project the incremental compilation should succeed.
2. The project compilation should succeed.

### Current Behavior
1. The compilation results in a StackOverFlowError
2. The compilation fails.

### Context
I'm trying to compile the project with the included annotation processors:
[annotation-processing-test.zip](https://github.com/gradle/gradle/files/3920953/annotation-processing-test.zip)
The project contains generated files in the `annottestgen` directory. These sources are compiled with the presence of the annotation processors in `processors.jar`. The source for the annotation processors are in `src_processors`, I included a compiled version of them in the JAR.

There are two annotation processors:

`ExtractProcessor`: That extracts an `interface` definition based on the annotated methods. (isolating)
`GroupProcessor`: Creates a group class that aggregates the annotated types. (aggregating)

There's also a commented out part in `testpackage0.TestClass1023` that removes a field definition. The `groups.Group7` class is generated by the `GroupProcessor` in the last round of annotation processing and an appropriate constructor is added to it. However, the compilation fails.


### Steps to Reproduce 
Stackoverflow:
I'm using the `path\to\gradle-6.0.1\bin\gradle.bat compileJava -i --stacktrace` command
1. Compile the project.
2. Add a new line at the start of the `testpackage0.TestClass1023` class.
3. Run the compilation again, and fail with the following stacktrace:

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileJava'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:187)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:185)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:166)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:374)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:361)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:354)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:340)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.StackOverflowError
        at com.google.common.collect.Sets$1.size(Sets.java:669)
        at com.google.common.collect.ImmutableSet.copyOf(ImmutableSet.java:251)
        at org.gradle.api.internal.tasks.compile.incremental.deps.DependentsSet.dependents(DependentsSet.java:37)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.getDependents(ClassSetAnalysis.java:169)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:151)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
...
```

Compilation error:
1. Uncomment the field definition in `testpackage0.TestClass1023`.
2. Compile the project, and fail:

```
annottestgen\testpackage0\TestClass1023.java:7: error: package groups does not exist
        groups.Group7 group = new groups.Group7(this);
              ^
```

It shouldn't, as the generated file at `build\generated\sources\annotationProcessor\java\main\groups\Group7.java` contains an appropriate constructor: `public Group7(testpackage0.TestClass1023 param) { } `.

See the attached archive at Context.

### Your Environment

```
------------------------------------------------------------
Gradle 6.0.1
------------------------------------------------------------

Build time:   2019-11-18 20:25:01 UTC
Revision:     fad121066a68c4701acd362daf4287a7c309a0f5

Kotlin:       1.3.50
Groovy:       2.5.8
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          1.8.0_221 (Oracle Corporation 25.221-b11)
OS:           Windows 10 10.0 amd64
```

The bugs are present in 5.1 and 5.6.4 as well.

The project compiles in 47 seconds when performing clean builds. (5 year old PC, 16GB memory, 4 cores)
I've tried 6.5 and 6.6-rc-1, both failing for incremental build.
The stacktrace is similar, I include it for 6.6-rc-1.

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileJava'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:208)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:206)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:187)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:409)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:399)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:94)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.execution.plan.DefaultPlanExecutor.process(DefaultPlanExecutor.java:72)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.executeWithServices(DefaultTaskExecutionGraph.java:184)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.execute(DefaultTaskExecutionGraph.java:177)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:39)
        at org.gradle.execution.DefaultBuildWorkExecutor.execute(DefaultBuildWorkExecutor.java:40)
        at org.gradle.execution.DefaultBuildWorkExecutor.access$000(DefaultBuildWorkExecutor.java:24)
        at org.gradle.execution.DefaultBuildWorkExecutor$1.proceed(DefaultBuildWorkExecutor.java:48)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:49)
        at org.gradle.execution.DefaultBuildWorkExecutor.execute(DefaultBuildWorkExecutor.java:40)
        at org.gradle.execution.DefaultBuildWorkExecutor.execute(DefaultBuildWorkExecutor.java:33)
        at org.gradle.execution.IncludedBuildLifecycleBuildWorkExecutor.execute(IncludedBuildLifecycleBuildWorkExecutor.java:36)
        at org.gradle.execution.DeprecateUndefinedBuildWorkExecutor.execute(DeprecateUndefinedBuildWorkExecutor.java:42)
        at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor$ExecuteTasks.run(BuildOperationFiringBuildWorkerExecutor.java:57)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:395)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:387)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:84)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor.execute(BuildOperationFiringBuildWorkerExecutor.java:42)
        at org.gradle.initialization.DefaultGradleLauncher.runWork(DefaultGradleLauncher.java:272)
        at org.gradle.initialization.DefaultGradleLauncher.doClassicBuildStages(DefaultGradleLauncher.java:179)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:155)
        at org.gradle.initialization.DefaultGradleLauncher.executeTasks(DefaultGradleLauncher.java:131)
        at org.gradle.internal.invocation.GradleBuildController$1.create(GradleBuildController.java:72)
        at org.gradle.internal.invocation.GradleBuildController$1.create(GradleBuildController.java:67)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:180)
        at org.gradle.internal.work.StopShieldingWorkerLeaseService.withLocks(StopShieldingWorkerLeaseService.java:40)
        at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:67)
        at org.gradle.internal.invocation.GradleBuildController.run(GradleBuildController.java:56)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:63)
        at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
        at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:52)
        at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:41)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:49)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:44)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:409)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:399)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:94)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:44)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:50)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:47)
        at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:87)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:47)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:31)
        at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:44)
        at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:29)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:87)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:55)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:60)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:38)
        at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:68)
        at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:38)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:37)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:26)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:60)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:32)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:56)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:42)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:48)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:32)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:68)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:82)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.StackOverflowError
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysisData.getDependents(ClassSetAnalysisData.java:56)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.getDependents(ClassSetAnalysis.java:160)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:151)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
        at org.gradle.api.internal.tasks.compile.incremental.deps.ClassSetAnalysis.processDependentClasses(ClassSetAnalysis.java:154)
```

```
------------------------------------------------------------
Gradle 6.6-rc-1
------------------------------------------------------------

Build time:   2020-07-13 13:53:25 UTC
Revision:     1ed79f4fbe18d90df8cb759235804f95a99b30a2

Kotlin:       1.3.72
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.8 compiled on May 10 2020
JVM:          1.8.0_221 (Oracle Corporation 25.221-b11)
OS:           Windows 10 10.0 amd64
```

The example project is significantly smaller than the [single-large-project](https://github.com/gradle/performance-comparisons/tree/master/single-large-project) Gradle benchmark project.

> There have been improvements to the incremental compiler and it shouldn't recurse anymore. 

May it have been that the infinite recursion got converted to an infinite loop? That 45 min sure didn't spend all time compiling. I recommend try taking a thread dump after a few minutes to see where it is stuck.

I've tried it with a smaller project (like 100 or 1000 source files), but the bug didn't happen that times. Not sure how it could relate to the size of the project though.

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java
@@ -18,16 +18,15 @@ package org.gradle.api.internal.tasks.compile.incremental.deps;
 
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSetMultimap;
-import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import it.unimi.dsi.fastutil.ints.IntSet;
 import org.gradle.api.internal.tasks.compile.incremental.processing.AnnotationProcessingData;
 import org.gradle.api.internal.tasks.compile.incremental.processing.GeneratedResource;
 
-import java.util.ArrayList;
+import java.util.ArrayDeque;
 import java.util.Collections;
+import java.util.Deque;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
@@ -140,37 +139,40 @@ public class ClassSetAnalysis {
                                          Set<GeneratedResource> resultResources,
                                          Iterable<String> privateDependentClasses,
                                          Iterable<String> accessibleDependentClasses) {
-        List<String> remainingAccessibleDependentClasses = Lists.newArrayList(accessibleDependentClasses);
-        List<String> remainingPrivateDependentClasses = Lists.newArrayList(privateDependentClasses);
-        while (remainingAccessibleDependentClasses != null) {
 
-            for (String d : remainingPrivateDependentClasses) {
-                if (!visitedClasses.add(d)) {
-                    continue;
-                }
-                privateResultClasses.add(d);
-                DependentsSet currentDependents = getDependents(d);
-                if (!currentDependents.isDependencyToAll()) {
-                    resultResources.addAll(currentDependents.getDependentResources());
-                }
+        for (String privateDependentClass : privateDependentClasses) {
+            if (!visitedClasses.add(privateDependentClass)) {
+                continue;
             }
-            remainingPrivateDependentClasses.clear();
+            privateResultClasses.add(privateDependentClass);
+            DependentsSet currentDependents = getDependents(privateDependentClass);
+            if (!currentDependents.isDependencyToAll()) {
+                resultResources.addAll(currentDependents.getDependentResources());
+            }
+        }
 
-            ArrayList<String> remainingAccessibleDependentClassesCopy = Lists.newArrayList(remainingAccessibleDependentClasses);
-            remainingAccessibleDependentClasses = null;
-            for (String d : remainingAccessibleDependentClassesCopy) {
-                if (!visitedClasses.add(d)) {
-                    continue;
-                }
-                accessibleResultClasses.add(d);
-                DependentsSet currentDependents = getDependents(d);
-                if (!currentDependents.isDependencyToAll()) {
-                    resultResources.addAll(currentDependents.getDependentResources());
-                    if (remainingAccessibleDependentClasses == null) {
-                        remainingAccessibleDependentClasses = Lists.newArrayList();
-                    }
-                    remainingAccessibleDependentClasses.addAll(currentDependents.getAccessibleDependentClasses());
-                }
+        processTransitiveDependentClasses(visitedClasses, accessibleResultClasses, resultResources, accessibleDependentClasses);
+    }
+
+    private void processTransitiveDependentClasses(Set<String> visitedClasses,
+                                                   Set<String> accessibleResultClasses,
+                                                   Set<GeneratedResource> resultResources,
+                                                   Iterable<String> accessibleDependentClasses) {
+        Deque<String> remainingAccessibleDependentClasses = new ArrayDeque<>();
+        for (String accessibleDependentClass : accessibleDependentClasses) {
+            remainingAccessibleDependentClasses.add(accessibleDependentClass);
+        }
+
+        while (!remainingAccessibleDependentClasses.isEmpty()) {
+            String accessibleDependentClass = remainingAccessibleDependentClasses.pop();
+            if (!visitedClasses.add(accessibleDependentClass)) {
+                continue;
+            }
+            accessibleResultClasses.add(accessibleDependentClass);
+            DependentsSet currentDependents = getDependents(accessibleDependentClass);
+            if (!currentDependents.isDependencyToAll()) {
+                resultResources.addAll(currentDependents.getDependentResources());
+                remainingAccessibleDependentClasses.addAll(currentDependents.getAccessibleDependentClasses());
             }
         }
     }

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java
@@ -18,13 +18,16 @@ package org.gradle.api.internal.tasks.compile.incremental.deps;
 
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import it.unimi.dsi.fastutil.ints.IntSet;
 import org.gradle.api.internal.tasks.compile.incremental.processing.AnnotationProcessingData;
 import org.gradle.api.internal.tasks.compile.incremental.processing.GeneratedResource;
 
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
@@ -131,27 +134,43 @@ public class ClassSetAnalysis {
      * further dependents, while resources are just data accumulated along the way. Recurses for classes that
      * are "publicly accessbile", i.e. classes that are not just used privately in a class.
      */
-    private void processDependentClasses(Set<String> visitedClasses, Set<String> privateResultClasses, Set<String> accessibleResultClasses, Set<GeneratedResource> resultResources, Iterable<String> privateDependentClasses, Iterable<String> accessibleDependentClasses) {
-        for (String d : privateDependentClasses) {
-            if (!visitedClasses.add(d)) {
-                continue;
-            }
-            privateResultClasses.add(d);
-            DependentsSet currentDependents = getDependents(d);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
-            }
-        }
+    private void processDependentClasses(Set<String> visitedClasses,
+                                         Set<String> privateResultClasses,
+                                         Set<String> accessibleResultClasses,
+                                         Set<GeneratedResource> resultResources,
+                                         Iterable<String> privateDependentClasses,
+                                         Iterable<String> accessibleDependentClasses) {
+        List<String> remainingAccessibleDependentClasses = Lists.newArrayList(accessibleDependentClasses);
+        List<String> remainingPrivateDependentClasses = Lists.newArrayList(privateDependentClasses);
+        while (remainingAccessibleDependentClasses != null) {
 
-        for (String d : accessibleDependentClasses) {
-            if (!visitedClasses.add(d)) {
-                continue;
+            for (String d : remainingPrivateDependentClasses) {
+                if (!visitedClasses.add(d)) {
+                    continue;
+                }
+                privateResultClasses.add(d);
+                DependentsSet currentDependents = getDependents(d);
+                if (!currentDependents.isDependencyToAll()) {
+                    resultResources.addAll(currentDependents.getDependentResources());
+                }
             }
-            accessibleResultClasses.add(d);
-            DependentsSet currentDependents = getDependents(d);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
-                processDependentClasses(visitedClasses, privateResultClasses, accessibleResultClasses, resultResources, Collections.emptySet(), currentDependents.getAccessibleDependentClasses());
+            remainingPrivateDependentClasses.clear();
+
+            ArrayList<String> remainingAccessibleDependentClassesCopy = Lists.newArrayList(remainingAccessibleDependentClasses);
+            remainingAccessibleDependentClasses = null;
+            for (String d : remainingAccessibleDependentClassesCopy) {
+                if (!visitedClasses.add(d)) {
+                    continue;
+                }
+                accessibleResultClasses.add(d);
+                DependentsSet currentDependents = getDependents(d);
+                if (!currentDependents.isDependencyToAll()) {
+                    resultResources.addAll(currentDependents.getDependentResources());
+                    if (remainingAccessibleDependentClasses == null) {
+                        remainingAccessibleDependentClasses = Lists.newArrayList();
+                    }
+                    remainingAccessibleDependentClasses.addAll(currentDependents.getAccessibleDependentClasses());
+                }
             }
         }
     }

